Expand test coverage for native AOT on android

sbomer opened the pull request:
- Enables building native AOT smoke tests for android with one APK per test project
- Adds marker files to set expected exit code when running through helix
- Adds a runtime-androidemulator job for native AOT smoke tests
- Expands libraries test coverage to include non-smoke tests
- Sets up appcontext base directory in test host
- Calls managed main through a pinvoke when running Exe project in the runtime tests

Fixes https://github.com/dotnet/runtime/issues/120715

sbomer committed:
Add more native AOT on android ci test coverage

- Expand native AOT smoke tests to all libraries tests - Add job for
runtime smoke tests using native AOT on emulator

sbomer committed:
Remove --cross argument, fix and disable some libs tests

sbomer committed:
Remove --cross argument, fix and disable some libs tests

sbomer committed:
Set runtimeFlavor to coreclr

sbomer committed:
Parameterize runtimeFlavor in Android job definitions

sbomer committed:
Remove runtimeFlavor: coreclr from Android emulator pipeline

sbomer committed:
Fix typo in runtimeFlavor parameter name

sbomer committed:
Add 'nativeaot' arg to testBuildArgs for Android emulator

sbomer committed:
Increase timeout to 180

sbomer committed:
Refactor Android NDK compiler selection for test builds

Move NDK compiler logic to a shared props file and import it in test and
build projects to unify Android NativeAOT and Mono builds.

sbomer committed:
Increase timeout to 240

sbomer committed:
Pass RuntimeFlavor to AndroidAppBuilderTask

sbomer committed:
Check TestBuildMode instead

sbomer committed:
Fix a few test failures

- Set ASSETS_DIR for crypto tests
  (using same pattern as existing crypto test)
- Set AppContext.BaseDirectory

We should probably change crypto tests to use
just AppContext.BaseDirectory.

sbomer committed:
Set Environment.CurrentDirectory to app files directory

sbomer committed:
Exclude tests failing due to missing calli marshalling support

sbomer committed:
Exclude tests failing due to missing crypto shim export

sbomer committed:
Enable building smoke tests as apks

- Use ReferenceXUnitWrapperGenerator != false to define SINGLE_FILE_TEST_RUNNER (android & singlefile targets) instead of IsFunctionalTest
- Condition SingleFileTestRunner compile and TrimmerRootAssembly (TestUtilities) on wrapper/generator and SkipTestUtilitiesReference
- Add Android NativeAOT publish/trim settings and set NativeLib/CustomNativeMain for android nativeaot
- Make DisableImplicitFrameworkReferences respect preexisting value and add explicit DisableImplicitFrameworkReferences=false in nativeaot tests dir
- Import testing\tests.targets when TargetsAndroid is true
- Update many NativeAOT smoke tests Main signatures to public static int Main(string[] args) to match runner invocation

sbomer committed:
Fix smoke test build for android

- Ensure they have Main entry point
- Tighten MSBuild conditions:
  - Only set NativeLib/CustomNativeMain for Android when OutputType == Exe
  - Only import testing/tests.targets for Android when using MicrosoftNETSdk and building executables
  - Make TestNativeAOT/UseNativeAOTRuntime defaults apply when TestBuildMode == nativeaot and only include monodroid-nativeaot.cs for executable/XUnit-wrapper cases

sbomer committed:
Disable directory apk build for smoke tests

sbomer committed:
Try using merged test runners

sbomer committed:
Revert "Try using merged test runners"

This reverts commit 4ed06ec5a6b6dd3254a32f4e5a1be32f80c9b7f5.

sbomer committed:
Add Merged marker for smoke tests

sbomer committed:
Try merged wrappers again

- Avoid importing logic for SharedLibrary
- Fix library filename
- Set HasMergedInTests to ensure projects build

sbomer committed:
Revert "Add Merged marker for smoke tests"

This reverts commit 37b7b4efa392b06a652756e439a6760974c266b1.

sbomer committed:
Fix some issues with non-standalone build

sbomer committed:
Remove buildAllTestsAsStandalone

sbomer committed:
Set expected exit code

sbomer committed:
Remove redundant import

This import is now covered by the IsMergedTestRunnerAssembly property.
Fixes a warning about a double import.
Fixes issue where mono runtime tests are getting incorrect assembly name
due to this undesired import.

sbomer committed:
Fix mono test build

sbomer committed:
Set exit code as metadata for helix

sbomer committed:
Write ExpectedExitCode.txt

This persists the expected exit code as a .txt file in the merged payload directory
so that it can flow to the helix xharness command.

sbomer committed:
Remove condition on monodroid-nativeaot.cs

To ensure it's still included in libs tests

sbomer committed:
Disable some failing tests

jkotas commented on src/libraries/tests.proj:
```diff
@@ -607,6 +607,24 @@
   </ItemGroup>
 
   <ItemGroup Condition="'$(TargetOS)' == 'android' and '$(TestNativeAOT)' == 'true'">
+    <!-- https://github.com/dotnet/runtime/issues/120722 -->
```
This should be fixed now

MichalStrehovsky commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostApplicationBuilderTests.cs:
```diff
@@ -231,7 +232,7 @@ public void CanConfigureAppConfigurationAndRetrieveFromDI(CreateBuilderFunc crea
         }
 
-        [Theory]
+        [ConditionalTheory]
```
What do these changes do?

sbomer committed:
Use __managed__Main, move suppession to common targets

sbomer committed:
Merge remote-tracking branch 'origin/main' into naot-android-coverage

sbomer committed:
Clean up disabled tests

- Use ActiveIssue
- Re-enable tests that are passing with AppContext.BaseDirectory fix
- Disable tests failing after merge
- Update issue link for System.Diagnostics.Debug.Tests

sbomer committed:
Revert SmokeTest entry point code changes

sbomer committed:
Skip/enable some more tests

sbomer commented on src/libraries/System.Runtime/tests/System.Reflection.Tests/AssemblyTests.cs:
```diff
@@ -149,6 +150,11 @@ public void ExportedTypes(Type type, bool expected)
         {
+            if (PlatformDetection.IsNativeAot && PlatformDetection.IsAndroid)
+            {
+                throw new SkipTestException("Disabled on NativeAOT Android: https://github.com/dotnet/runtime/issues/121209");
```
The failure is NativeAOT only - I'll keep using that issue to track this since it was reopened and has some more detail now.

sbomer committed:
Fix non-android build, disable test

sbomer commented:
/azp run runtime-androidemulator
/azp run runtime-android

sbomer committed:
Re-enable passing tests

sbomer commented:
/azp run runtime-androidemulator
/azp run runtime-android

sbomer committed:
PR feedback

- Clean up whitespace diffs
- Clean up TEST_ASSETS reference
- ifdef out test logic that fails on Android
- Work around failure by not using libc++

sbomer commented:
/azp run runtime-android

sbomer committed:
PR feedback and cleanup

- Clean up ReferenceXUnitWrapperGenerator condition
- Disable tests for android via runtime check
- Disable unsupported ISA tests
- Add comment for disabled SharedLibrary test

sbomer commented on src/tests/nativeaot/Directory.Build.props:
```diff
@@ -6,4 +6,14 @@
+    <HasMergedInTests>true</HasMergedInTests>
+    <ExpectedExitCode>100</ExpectedExitCode>
+    <DefineConstants>$(DefineConstants);TARGETS_ANDROID</DefineConstants>
+  </PropertyGroup>
```
I haven't looked at the rest of the runtime tests yet, not sure how much they will differ - probably some but not all of this will need to be shared (added a comment).

sbomer replied:
I see, moved to the parent props.

sbomer committed:
Add comment about runtime tests

sbomer committed:
Try to fix condition for SingleFileTestRunner, fix test failure

sbomer committed:
Move shared props out of nativeaot dir

And fix resulting duplicate imports
due to build.proj getting these settings.

sbomer committed:
Fix build error, disable test, and try to clean up condition

sbomer committed:
Adjust condition

sbomer committed:
Disable ThreadTest

sbomer commented:
/azp run runtime-androidemulator

MichalStrehovsky commented on src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostApplicationBuilderTests.cs:
```diff
@@ -400,7 +400,7 @@ private static string EnsureTrailingSlash(string path)
                 return path + Path.DirectorySeparatorChar;
             }
- 
+
```
Did you mean to leave this in?

MichalStrehovsky reviewed (approved):
Looks good but I don't know anything about xharness or android testing. Would be nice if someone else who knows about those could look at it from that angle.

sbomer committed:
PR feedback, increase timeout

- Fix left-over whitespace change
- Remove unnecessary ClrTestTargetUnsupported condition

sbomer commented:
/azp run runtime-androidemulator

sbomer commented:
/azp run runtime-nativeaot-outerloop

jtschuster reviewed (approved):
Looks good to me!

sbomer committed:
Consolidate NdkToolchainPrebuiltOS, disable tests

sbomer commented:
/azp run runtime-androidemulator

sbomer commented:
/azp run runtime-android
